### DISCUSSION TITLE
Lua loading issue

### DISCUSSION THREAD

**Discussion by xsh005** [lumen-oss/rocks.nvim#122] [Q&A] [answered]
The problem is somewhat hard to describe, so I am going to try my best.
My `~/.config/nvim/after/plugin` contains some configurations for plugins
```
dap.lua        
epo.lua        
guard.lua      
lspconfig.lua  
squirrel.lua   
telescope.lua  
treesitter.lua 
```
If I open nvim in that particular directory, I would get warnings like these:
```
E5113: Error while calling lua chunk: ./dap.lua:1: loop or previous error loading module 'dap'
```
dap.lua:
```lua
1 local dap = require("dap")
```
I suspect the lua loader module is confusing the local dap file with the dap directory in my rtp, causing an infinite require cycle.

**Comment by mrcjkb** [comment]
Hey 👋 

I don't think this is related to rocks.nvim.
We only add plugins to the runtimepath, but don't source `ftplugin` directories. That's left up to Neovim.
And afaik, Lua's package loader won't search for modules in `plugin` or `ftplugin` directories.

You shouldn't put those files in
`after/ftplugin`, which is meant for filetype-specific scripts (named after the filetype, e.g. `after/ftplugin/markdown.lua` for anything you want to be sourced when you open a markdown file).

Your plugin configs can go in `~/.config/nvim/plugin` instead.

Regarding the error message you're seeing: That usually happens when you have cyclic module dependencies.

**Reply by xsh005** [reply]
Oops, typo!
I meant `after/plugin` :) `rocks.nvim` do source these, right?
As for cyclic dependency, this also happens for `squirrel.lua`, which literally just contains `require("squirrel")`
Sorry for the confusion :(

**Reply by mrcjkb** [reply]
rocks.nvim only sources files in the *plugins'* `plugin` and `ftdetect` directories, but not the ones in your *user directory*. That's done by Neovim.

Do you observe the same behaviour when installing the plugins into the packpath manually, or with rocks-git.nvim?

**Reply by xsh005** [reply]
Hmm, I will try to come up with a minimal repro I guess!

**Reply by mrcjkb** [reply]
I just tried with both `nvim-dap` and `squirrel.nvim`, and couldn't reproduce it :disappointed: 
Do you have a link to your dotfiles? Maybe I can find something...

**Reply by xsh005** [reply]
Wow, thanks for the effort!
My dotfiles are [here](https://github.com/xiaoshihou514/nvim)
I do still get loading errors when nvim starts in `after/plugin`

**Reply by mrcjkb** [reply]
> when nvim starts in after/plugin

Wait, do I understand this correctly? The errors occur when you open one of the files in `after/plugin`, not when you open Neovim?
:thinking: I can't seem to be able to reproduce that either.
Will have a look at your config :eyes:

**Reply by mrcjkb** [reply]
I just cloned your repo into `~/.config/nanovim` and tried to reproduce it with `NVIM_APPNAME=nanovim`.

- Modified `vim.g.rocks_nvim` to use my home directory
- `:Rocks sync` -> Installed everything, except for the git plugins
- Restarted
- `:Rocks sync` -> Installed all the git plugins
- Restarted -> No errors
- Restarted in `~/.config/after/plugin` -> No errors
- Restarted in `~/.config/after/plugin/squirrel.lua` (nice plugin btw :smile:) -> No errors

My wild guess is that somehow your plugin installations (either in the rocks install directory or the rocks-git install directory) got messed up?

I would suggest you try uninstalling (`:Rocks prune` or comment out, then `:Rocks sync`), and then reinstalling the affected plugins.

**Reply by mrcjkb** [reply, marked as answer]
By the way, we also have a [rocks-config.nvim](https://github.com/nvim-neorocks/rocks-config.nvim) module, which helps you safely load your plugins (skipping ones that aren't installed). Maybe that can help in your case?

**Reply by xsh005** [reply]
> Wait, do I understand this correctly? The errors occur when you open one of the files in `after/plugin`, not when you open Neovim? 🤔 I can't seem to be able to reproduce that either. Will have a look at your config 👀

It happens when I open nvim in the `~/.config/nvim/after/plugin` directory (in your case `~/.config/nanovim/after/plugin`), but it seems that it's not really reproducible lol.
~I will try to `:Rocks prune` and see if it goes away.~ Nope :(

**Reply by xsh005** [reply]
Using `rocks-config.nvim` does solve the issue, but the issue only happens in that particular directory anyway. I was hoping I could catch some potential loading issues there. :thinking:

**Reply by xsh005** [reply]
I guess it''s not worth it to spend so much time into a niche issue that can't be reproduced :) Closing then.

**Reply by mrcjkb** [reply]
:thinking: maybe manually removing everything from the install directories and packpath. and then `:Rocks sync`ing could help?
Just be careful not to remove rocks.nvim and its dependencies...
